earth2studio/models/auto.py

```diff
@@ -14,85 +14,287 @@
+            fn = os.path.join(self.storage[-1], hash)
+            user_specified_kwargs = {
+                k: v
+                for k, v in kwargs.items()
+                # those kwargs were added by open(), we don't want them
+                if k not in ["autocommit", "block_size", "cache_options"]
+            }
+            return LocalTempFile(self, path, mode=mode, fn=fn, **user_specified_kwargs)
+        detail = self._check_file(path)
+        if detail:
+            detail, fn = detail
+            _, blocks = detail["fn"], detail["blocks"]
+            if blocks is True:
+                f = open(fn, mode)
+                f.original = detail.get("original")
+                return f
+            else:
+                raise ValueError(
+                    f"Attempt to open partially cached file {path}"
+                    f" as a wholly cached file"
+                )
+        else:
+            fn = self._make_local_details(path)
+        kwargs["mode"] = mode
+
+        # call target filesystems open
+        self._mkcache()
+        if self.compression:
+            with self.fs._open(path, **kwargs) as f, open(fn, "wb") as f2:
+                if isinstance(f, AbstractBufferedFile):
+                    # want no type of caching if just downloading whole thing
+                    f.cache = BaseCache(0, f.cache.fetcher, f.size)
+                comp = (
+                    infer_compression(path)
+                    if self.compression == "infer"
+                    else self.compression
+                )
+                f = compr[comp](f, mode="rb")
+                data = True
+                while data:
+                    block = getattr(f, "blocksize", 5 * 2**20)
+                    data = f.read(block)
+                    f2.write(data)  # type: ignore
+        else:
+            if "callback" in kwargs:  # Patch here
+                self.fs.get_file(path, fn, callback=kwargs["callback"])
+            else:
+                self.fs.get_file(path, fn)
+        self.save_cache()
+        return self._open(path, mode)  # type: ignore
+
+
+class TqdmFormat(tqdm):
+    """Provides a `total_time` format parameter. Not used.
+    See: https://filesystem-spec.readthedocs.io/en/stable/api.html#fsspec.callbacks.TqdmCallback
+    """
+
+    @property
+    def format_dict(self) -> dict:
+        d = super().format_dict
+        return d
+
+
+class TqdmCallbackRelative(TqdmCallback):
+    """Simple extention of Tqdm callback to support progress bar on recurrive gets"""
+
+    def branched(self, path_1, path_2, **kwargs) -> Callback:  # type: ignore
+        """Child callback for recursive get"""
+        tqdm_kwargs = self._tqdm_kwargs
+        tqdm_kwargs["unit"] = "B"
+        tqdm_kwargs["unit_scale"] = True
+        tqdm_kwargs["unit_divisor"] = 1024
+        callback = TqdmCallback(
+            tqdm_kwargs=tqdm_kwargs,
+            tqdm_cls=self._tqdm_cls,
+        )
+        return callback
```

Comment:
This is going to sound very small but I think we can find a better place to put these utility classes.

Reply:
I agree if this gets used in more spots. Would be up for moving it once another part of the code base uses it (like a file system util class). Similar to the onnx utils. I'm not sure this will be used in the data sources, with the current refactoring of them I'm even less sure...

Reply:
Honestly, I'm hoping a fix gets upstreams to fsspec and eventually the cache class can get removed.